scipy/stats/_unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,1122 @@
+            unur_set_stream(self._messages.handle)
+
+            for i in range(size):
+                out[i] = unur_sample_discr(rng)
+                if PyErr_Occurred():
+                    return
+
+            msg = self._messages.get()
+            if msg:
+                raise UNURANError(msg)
+        finally:
+            _lock.release()
+            if self._callback_wrapper is not None:
+                release_unuran_callback(&callback)
+
+    def rvs(self, size=None):
+        """
+        Sample from the distribution.
+
+        Parameters
+        ----------
+        size : int or tuple, optional
+            The shape of samples. Default is ``None`` in which case a scalar
+            sample is returned.
+
+        Returns
+        -------
+        rvs : array_like
+            A NumPy array of random variates.
+        """
+        cdef double[::1] out_cont
+        cdef int[::1] out_discr
+        N = 1 if size is None else np.prod(size)
+        if unur_distr_is_cont(unur_get_distr(self.rng)):
+            out_cont = np.empty(N, dtype=np.float64)
+            self._rvs_cont(out_cont)
+            if size is None:
+                return out_cont[0]
+            return np.asarray(out_cont).reshape(size)
+        elif unur_distr_is_discr(unur_get_distr(self.rng)):
+            out_discr = np.empty(N, dtype=np.int32)
+            self._rvs_discr(out_discr)
+            if size is None:
+                return out_discr[0]
+            return np.asarray(out_discr).reshape(size)
+        else:
+            raise NotImplementedError("only univariate continuous and "
+                                      "discrete distributions supported")
+
+    @property
+    def seed(self):
+        return self.numpy_rng
+
+    @seed.setter
+    def seed(self, new_seed):
+        self.numpy_rng = get_numpy_rng(new_seed)
+        _lock.acquire()
+        try:
+            self._messages.clear()
+            unur_set_stream(self._messages.handle)
+            unur_urng_free(self.urng)
+            self._urng_builder = _URNG(self.numpy_rng)
+            self.urng = self._urng_builder.get_urng()
+            if self.urng == NULL:
+                raise UNURANError(self._messages.get())
+            unur_chg_urng(self.rng, self.urng)
+        finally:
+            _lock.release()
+
+    @cython.final
+    def __dealloc__(self):
+        if self.distr != NULL:
+            unur_distr_free(self.distr)
+            self.distr = NULL
+        if self.par != NULL:
+            unur_par_free(self.par)
+            self.par = NULL
+        if self.rng != NULL:
+            unur_free(self.rng)
+            self.rng = NULL
+        if self.urng != NULL:
+            unur_urng_free(self.urng)
+            self.urng = NULL
+
+    # Pickling support
+    @cython.final
+    def __reduce__(self):
+        klass = functools.partial(self.__class__, **self._kwargs)
+        return (klass, ())
+
+
+cdef class TransformedDensityRejection(Method):
```

Comment:
The documentation does not show the signature of the class initializer or link to the source:
![image](https://user-images.githubusercontent.com/6570539/128557806-3820751a-cb9e-404f-867e-6503b5ffc308.png)

as it does for Python classes, e.g.
![image](https://user-images.githubusercontent.com/6570539/128557866-953bb076-b9bf-4c29-9ce0-69c004eb4d7a.png)

The same is true of the methods. I don't think I can tell from the documentation, for example, which (if any) arguments are keyword-only. Can we get around this?

Can we get around this?

Reply:
> Can we get around this?

I can't find anything in Cython docs. NumPy explicitly states the signatures in the doc string. e.g [Philox BitGenerator](https://github.com/numpy/numpy/blob/261a769dd135c1e4007a6d0bdbed68a8c929497a/numpy/random/_philox.pyx#L56):

```py
cdef class Philox(BitGenerator):
    """
    Philox(seed=None, counter=None, key=None)

    ...
```

and it looks like this on the doc pages:

![image](https://user-images.githubusercontent.com/43181252/128586075-6ef85249-fbd5-4f2b-915a-ec9675cc0415.png)

I think this addresses the problem nicely. I will push this and we can see how it looks.

Reply:
Done.

![image](https://user-images.githubusercontent.com/43181252/128597705-92ffdd0b-ec13-4a99-9a2c-692f2942f69c.png)

Reply:
OK. If you happen to figure out how to manually specify the "[source]" link, it would be great to know. But I'll consider this resolved.

Reply:
> OK. If you happen to figure out how to manually specify the "[source]" link, it would be great to know. But I'll consider this resolved.

Hmm. I didn't think about the source button. I am not sure how to do that: NumPy also doesn't have that for Cython extension classes.